### ISSUE TITLE
area_grad does not return unit vectors

### ISSUE THREAD

**Issue by btbest** [DamCB/tyssue#237]
Since I'm trying to get the viscous implementation working, I've been digging deeper into the gradient functions, which landed me at `dynamics.sheet_gradients.area_grad` and `.height_grad`.

For the `settings["geometry"]` terms included in tyssue, height_grad seems to return vectors of magnitude 1, oriented along the vertices' individual height axes, which are then mulitplied by the force in the respective effectors' `gradient()` function to obtain the force vector at the correct orientation and magnitude - makes perfect sense.

Reading the code for area_grad, it seems to try to achieve the same thing, and as far as I can tell, it really should. However, `np.linalg.norm( area_grad(sheet)[0], axis=1)` shows that both grad_a_srce and grad_a_trgt vectors returned by area_grad are on average of magnitude 1.63 in my case (the ellipsoid sheet from @sophietheis paper).

I confirmed that it's not just in my diagnostic setup but in my actual simulations by adding 
```
    print(np.linalg.norm(grad_a_srce,axis=1).mean())
    print(np.linalg.norm(grad_a_trgt,axis=1).mean())
```
before the function's return statement as well, which resulted in:
```
1.6257241576537274
1.6257241576537271
1.6252941611195968
1.625294161119597
1.6256257201234106
1.6256257201234106
1.6236565846003965
1.623656584600396
1.624853955149997
#etc
```

Now before I try figuring out why this isn't giving unit vectors: Am I wrong, and this is perfectly expected output? It seemed to me that there is something wrong here because the effectors then multiply the area_grad return value with the force, which in my case must be upscaling all the FaceAreaElasticity forces by that factor of 1.63... I'd be grateful for some quick feedback on whether this is expected behaviour or not; if it's buggy, I will report back if I find out what's wrong :)

**Comment by glyg** [comment]
Hi @btbest I think this is expected.

For each vertex, the gradient of the area at that vertex is the sum of the area gradients from the triangles formed by this vertex, its neighbors (linked by a half edge)  and the face centers. 

 
If we look at coordinate $x_i$ of the vertex i, and the triangle formed by the half-edge  $(i, j)$ and face $\alpha$, the gradient term is:

![Capture d’écran de 2021-04-30 09-23-51](https://user-images.githubusercontent.com/1426279/116662281-dd1fc480-a995-11eb-80bb-1f1c370987d2.png)

This as no reason to be normed,

Hope this helps,

Best
Guillaume

**Comment by btbest** [comment]
What you describe, and the variable names you use, sounds like the calculation for edge_df["sub_area"]. `sheet.edge_df[sheet.ncoords] = np.cross(r_ai, r_ij)` is in `sheet_geometry.update_normals`, 
and `sheet.edge_df["sub_area"] = (
            np.linalg.norm(sheet.edge_df[sheet.ncoords], axis=1) / 2
        )` is in `sheet_geometry.update_areas`

What the function area_grad calculates is:
```
    inv_area = sheet.edge_df.eval("1 / (4 * sub_area)")

    face_pos = sheet.edge_df[["f" + c for c in coords]].values
    srce_pos = sheet.edge_df[["s" + c for c in coords]].values
    trgt_pos = sheet.edge_df[["t" + c for c in coords]].values

    r_ak = srce_pos - face_pos # vector from centroid to srce
    r_aj = trgt_pos - face_pos

    inv_area = to_nd(inv_area, 3)
    grad_a_srce = inv_area * np.cross(r_aj, sheet.edge_df[ncoords])
    grad_a_trgt = inv_area * np.cross(sheet.edge_df[ncoords], r_ak)
```

Now that you say it, `np.cross(r_aj, sheet.edge_df[ncoords])` would indeed have to be divided not only by sub_area in order to obtain a vector of integer magnitude, so I jumped to a conclusion there while trying to figure out what this function is supposed to achieve. But it also isn't clear to me why the vector is divided by sub_area*4 either...

I am happy to take from this though that the function is working as intended and leave it at that, just wanted to make sure this isn't the reason why my viscous implementation isn't working (I'm getting happy porcupines over here!)

![grafik](https://user-images.githubusercontent.com/63287233/116700656-5257bd80-a9c7-11eb-8bc5-25e759244e51.png)

**Comment by glyg** [comment]
@sophietheis has a collection of such model failures, I'm sure she'll be happy to add this one (with your permission)

So, the term I wrote above is the differential with respect to x of the area of the triangle (so the sub_area). Due to [differentiation rules](https://en.wikipedia.org/wiki/Vector_calculus_identities#Dot_product_rule) you can express that differential as 
(r_aj x n_ij) / 4 A_aij - what's caclulated in the `area_grad` method (there is a calculus for when the vector is at the source of the half-edges, and one for when it's a target, for a 3-cells junction, that's 6 triangles:

![triangles](https://user-images.githubusercontent.com/1426279/116704980-53d7b480-a9cc-11eb-8831-8509dc92a3ed.png)

As for you crash, have you tried a smaller step? It is very possible that the area gradient computation is numerically unstable for small sizes. I also think in the Euler solver you can limit the maximum displacement, maybe that can help

**Comment by btbest** [comment]
I would be honoured to be included in the hall of fame/failures :"D

I did some geometry and figured that if I wanted that function to return unit vectors, it would need to divide by 2 * sub_area * norm(r_aj) or norm(r_ak) respectively. So I tried that, confirmed that it now gives unit vectors (which it does), ran the sim again, and the porcupine spikes are now distributed a bit differently but basically the same result.

So that's that chapter closed - thanks for the explanations!
